plugins/org.python.pydev.parser/src/org/python/pydev/parser/grammar310/TreeBuilder310.java

```diff
@@ -632,76 +634,250 @@ public final SimpleNode onCloseNode(SimpleNode n, int arity) throws Exception {
                     }
                     popNode = stack.popNode();
-                    try {
-                        pattern = (exprType) popNode;
-                    } catch (Exception e) {
-                        Log.log("Expected expr. Found: " + popNode);
+                    if (popNode instanceof patternType) {
+                        pattern = (patternType) popNode;
+                    } else if (popNode instanceof exprType) {
+                        pattern = new MatchValue((exprType) popNode);
                     }
+                    return new MatchAs(pattern, name);
                 } else {
-                    Log.log("Expected arity to be 2 here.");
+                    Log.log("Expected arity to be 1 or 2 here.");
                 }
-                return new MatchAs(pattern, name);
+                return null;
 
-            case JJTOR_PATTERN:
-                MatchOr matchOr = null;
-                if (arity > 1) {
-                    exprType[] patterns = new exprType[arity];
+            case JJTSTAR_PATTERN:
+                if (arity == 1) {
+                    SimpleNode popNode = stack.popNode();
+                    String name = NodeUtils.getFullRepresentationString(popNode);
+                    return new MatchStar(name);
+                } else {
+                    Log.log("Expected arity to be 1 here.");
+                }
+                return null;
+
+            case JJTOPEN_SEQUENCE_PATTERN:
+                if (arity > 0) {
+                    patternType[] patterns = new patternType[arity];
                     for (int i = 0; i < arity; i++) {
                         SimpleNode popNode = stack.popNode();
-                        try {
-                            patterns[i] = (exprType) popNode;
-                        } catch (Exception e) {
-                            Log.log("Expected expr. Found: " + popNode);
+                        if (popNode instanceof patternType) {
+                            patterns[i] = (patternType) popNode;
+                        } else if (popNode instanceof exprType) {
+                            patterns[i] = new MatchValue((exprType) popNode);
                         }
                     }
-                    matchOr = new MatchOr(patterns);
+                    return new MatchSequence(patterns);
+                } else if (arity == 1) {
+                    SimpleNode popNode = stack.popNode();
+                    if (popNode instanceof patternType) {
+                        return popNode;
+                    } else if (popNode instanceof exprType) {
+                        return new MatchValue((exprType) popNode);
+                    }
                 } else {
-                    Log.log("Expected arity to be > 1 here.");
+                    Log.log("Expected arity to be > 0 here.");
                 }
-                return matchOr;
+                return null;
+
+            case JJTGUARD:
```

Comment:
This doesn't seem to solve any purpose, so, it could probably map to `#void` instead of generating a node.